### ISSUE TITLE
Android GStreamer pipeline only starts once at application startup

### ISSUE THREAD

**Issue by Frequency21** [mavlink/qgroundcontrol#11676]
Disclaimer: This is more probably a GStreamer bug than a QGroundControl. But I'm hoping others ran into this issue and can share their experience or shred some light how to resolve this.

## Expected Behavior
Video stream should start every time, when RTSP URL changes or stream type changes.

## Current Behavior
When QGC is preconfigured with an IP camera's RTSP URL, then it can start displaying it. But after that, if there's a change in the URL, or chaning the video source to disabled then RTSP again, then the display is blank (fully black).

This is occurs only with a specific camera. [Topotek KHP290G609L11](https://topotek.com/KHP290G609L11-1080P-Day-light-640-Thermal1100m-LRF-Triple-Light-Gimbal-HDMIIP-output-p4777857.html)

## Steps to Reproduce:
Please provide an unambiguous set of steps to reproduce the current behavior
1. Install fresh QGC v4.3.0 or v4.4.0 on an android device (SkyDroid H16 in my case)
2. Set the video source to RTSP at general settings and configure the RTSP URL as well.
3. Exit the settings.
4. Video display is blank.

## System Information
When posting bug reports, include the following information
- Operating System: Android 7.1.2
- QGC Version: 4.4.0
- QGC build: stable
- Flight Controller: Matek H743
- Autopilot (with version): ArduCopter v4.5.0

## Detailed Description
I also tried to change HW decoder to software decoder (it renders a couple of frame every time the watchdog restarts the pipeline, but then it freezes). Increase the rank of another HW decoder (omx). Change the RTSP stream codec from H265 to H264.

For me (a total gstreamer newbie) it looks like the decoder can not negotiate with the upstream elements (the parser elements inside decodebin3, or if I replace the decodebin3 with the HW decoder element, then it cannot negotiate with the parsebin element). This is because parserbin and decodebin3 uses a typefind GSTElement, which guesses the stream format wrong.

Preconfigured startup (see attached logs.txt at line 2652):
> [I] at ../plugins/elements/gsttypefindelement.c:181 - "<typefind> found caps video/x-h265, width=(int)1920, height=(int)1080, framerate=(fraction)0/1, parsed=(boolean)true, stream-format=(string)byte-stream, alignment=(string)au, profile=(string)main, chroma-format=(string)4:2:0, bit-depth-luma=(uint)8, bit-depth-chroma=(uint)8, colorimetry=(string)1:3:5:1, tier=(string)main, level=(string)4.1, probability=100"

After URL changes see how the width, height and profile are incorrect (line 7108):
> [I] at ../plugins/elements/gsttypefindelement.c:181 - "<typefind> found caps video/x-h265, width=(int)16, height=(int)16, framerate=(fraction)0/1, parsed=(boolean)true, stream-format=(string)byte-stream, alignment=(string)au, profile=(string)main-10, probability=100"

The video decoder also changes profile (line 2974):
> [I] at ../gst-libs/gst/video/gstvideodecoder.c:1448 - "<amcvideodec-omxarmvideov5xxdecoder0> upstream tags: taglist, video-codec=(string)"H.265\ \(Main\ Profile\)";"

After (line 7259):
> [I] at ../gst-libs/gst/video/gstvideodecoder.c:1448 - "<amcvideodec-omxarmvideov5xxdecoder1> upstream tags: taglist, video-codec=(string)"H.265\ \(Main\ 10\ Profile\)";"

## Log Files and Screenshots
- Log file: [logs.txt](https://github.com/user-attachments/files/16193286/logs.txt)
- Pipelines after videosink is linked.
  - When video starts correctly on application startup: ![working-pipeline-with-videosink](https://github.com/user-attachments/assets/b8c01984-cacb-47aa-8701-6c466aba06ed)
  - When URL is changed and decoding stops: ![not-working-pipeline-with-videosink](https://github.com/user-attachments/assets/f5ee378d-28c7-479e-99a4-f84772154be0)
- Please notice how similar the two pipeline except inside decodebin3 capsfilter element.
- I also found a similar issue [#9440](https://github.com/mavlink/qgroundcontrol/issues/9440#issuecomment-787417912) but that case the reporter couldn't retest the application with the same camera.

**Comment by Viliuks** [comment]
@Frequency21 any updates on this issue? I'm encountering the same thing but on android 9, qgc 4.4.3, gstreamer 1.18.6

**Comment by Viliuks** [comment]
@HTRamsey any updates on this issue? Do you need any other information?

**Comment by HTRamsey** [comment]
No, it's definitely something I should prioritize but I just haven't yet. I know Don is planning to make an effort at video stuff sometime soon and hopefully I'll join in and fix some of this stuff by then

**Comment by Viliuks** [comment]
> No, it's definitely something I should prioritize but I just haven't yet. I know Don is planning to make an effort at video stuff sometime soon and hopefully I'll join in and fix some of this stuff by then

Let me know if you need any help, I would be more than willing to help you guys out!

**Comment by DonLakeFlyer** [comment]
Just to clarify:
* You are starting QGC and the Video settings start out at disabled
* You then change to RTSP and type in a valid URL

Questions:
* Please try this with Stable v5, 4.4 is no longer supported
* After settings RTSP URL exit QGC (Q Icon, Close Application) and then restart QGC and see what happens

**Comment by Frequency21** [comment]
Hello! Yes you're right.

To clarify: I started QGC with disabled video settings. Then I changed to RTSP with a valid URL, now the display is blank. (Your second question) After a restart the video stream is displayed without any settings modification.

Your first question: I tried other stable releases, ~~including v5~~ (or v4.8 more likely? I have faint memories), but it behaved the same way.

Unfortunately I don't have the hardware anymore to make tests.

**Comment by DonLakeFlyer** [comment]
@Viliuks Do you see the same proble, with 5.0? If so please provide exact steps to reproduce

**Comment by Viliuks** [comment]
> [@Viliuks](https://github.com/Viliuks) Do you see the same proble, with 5.0? If so please provide exact steps to reproduce

Sorry for the late response,
but it happens once you try to update the RTSP url or the camera drops the connection. Then the pipeline isnt able to be brought back. 
But it seems to happen with specific cameras (havent checked the handshake packets), but when you are using a UDP source this issue doesnt occur and you are able to reuse the same videoreceiver for different UDP sources.

What I have been able to do, is switch from android hw decoder to sw decoder without using the decodebin element and it mitagates the issue.

What I have found, that it seems there is a problem with the `decodebin3` gstreamer element, where it cannot create the required caps and it bricks the pipeline with 1x1 image size and the downstream elements fail.

Code snippet to mitigate the issue.
```
  if (!_uri.isEmpty() && (_uri.contains(":554/22", Qt::CaseInsensitive)))
        {
            GstElement* parser     = gst_element_factory_make("h265parse",   "h265parse_rtsp");
            GstElement* swdecoder  = gst_element_factory_make("avdec_h265",  "avdec_h265_rtsp");
            GstElement* postQueue  = gst_element_factory_make("queue",       "queue_post_decoder_rtsp");
...
else if ((decoder = gst_element_factory_make("decodebin3", nullptr)) == nullptr) {
            qCCritical(VideoReceiverLog) << "gst_element_factory_make('decodebin3') failed";
            break;
        }
```

Also some notes, I have noticed this issue on 4.3.3 and 5.0 versions. And also, I do not really have access to this type of camera anymore, so I cant really test it. (I might be able to get access if need be, but I have since moved on to a different sensor and payload).

Also these issues are experienced when you are running QGC on Android.

Generally, I think its an issue of object lifespan during QGC app lifecycle*. As in, if I create two different videoreceivers and I do not change anything, they run correctly without an issue. You can stop/play the videostream and it works just fine. But once you change anything in the receiver end then it bricks the downstream pipeline. So the general solution I have came up with is just to use multiple receivers, for multiple cameras and just change the QT widget drawing order for the gstreamer QT video sinks.

What might be the correct solution is to renegotiate the full gstreamer pipeline instead of just the source elements.

*note: I havent had the time to debug the lifespan of objects or gstreamer elements.
**note: Also its highly probable, that the cause is HW decoders of Qualcomm (QV450) instead of the camera itself. As the decodebin might be sending different caps than the HW decoder is expecting and the hw decoder just denies them and silently fails.

**Comment by DonLakeFlyer** [comment]
Can folks who are able to reproduce this test on latest daily just for the sake of sanity?

**Comment by DonLakeFlyer** [comment]
Using daily I switched from an RTSP video stream and then to a UDP video stream and it worked fine. Re-open if you can reproduce on latest daily.
